lib/config.ml

```diff
@@ -81,46 +107,72 @@ module Ciphers = struct
 
   let fs = fs_of default
 
+  let psk_of = List.filter Ciphersuite.ciphersuite_psk
+
+  let default13 = [
+    `TLS_AES_128_GCM_SHA256 ;
+    `TLS_AES_256_GCM_SHA384 ;
+    (* `TLS_CHACHA20_POLY1305_SHA256 ; *)
+    `TLS_AES_128_CCM_SHA256 ;
+    (* `TLS_AES_128_CCM_8_SHA256 *)
+  ]
+
+  let supported13 = default13
+
 end
 
-let default_hashes =
-  [ `SHA512 ; `SHA384 ; `SHA256 ; `SHA224 ; `SHA1 ]
+(* TODO split into <=12 and >=13, the SHA1 isn't 13 anymore *)
+let default_signature_algorithms =
+  [ `RSA_PSS_RSAENC_SHA256 ;
+    `RSA_PSS_RSAENC_SHA384 ;
+    `RSA_PSS_RSAENC_SHA512 ;
+    `RSA_PKCS1_SHA256 ;
+    `RSA_PKCS1_SHA384 ;
+    `RSA_PKCS1_SHA512 ;
+    `RSA_PKCS1_SHA1 ]
 
-let supported_hashes =
-  default_hashes @ [ `MD5 ]
+let supported_signature_algorithms =
+  default_signature_algorithms @ [ `RSA_PKCS1_MD5 ]
 
 let min_dh_size = 1024
 
 let min_rsa_key_size = 1024
 
-let dh_group = Dh.Group.ffdhe2048 (* ff-dhe draft 2048-bit group *)
+let dh_group = `FFDHE2048 (* ff-dhe draft 2048-bit group *)
+
+let supported_groups =
+  [ `X25519 ; `P256 ; `FFDHE2048 ; `FFDHE3072 ; `FFDHE4096 ; `FFDHE6144 ; `FFDHE8192 ]
 
 let default_config = {
-  ciphers           = Ciphers.default ;
-  protocol_versions = (TLS_1_0, TLS_1_2) ;
-  hashes            = default_hashes ;
-  use_reneg         = false ;
-  authenticator     = None ;
-  peer_name         = None ;
-  own_certificates  = `None ;
-  acceptable_cas    = [] ;
-  session_cache     = (fun _ -> None) ;
-  cached_session    = None ;
-  alpn_protocols    = [] ;
+  ciphers = Ciphers.default ;
+  ciphers13 = Ciphers.default13 ;
+  protocol_versions = (TLS_1_0, TLS_1_3) ;
+  signature_algorithms = default_signature_algorithms ;
+  use_reneg = false ;
+  authenticator = None ;
+  peer_name = None ;
+  own_certificates = `None ;
+  acceptable_cas = [] ;
+  session_cache = (fun _ -> None) ;
+  cached_session = None ;
+  cached_ticket = None ;
+  alpn_protocols = [] ;
+  groups = supported_groups ;
+  ticket_cache = None ;
+  zero_rtt = 0l ;
 }
 
 let invalid msg = invalid_arg ("Tls.Config: invalid configuration: " ^ msg)
 
 let validate_common config =
   let (v_min, v_max) = config.protocol_versions in
   if v_max < v_min then invalid "bad version range" ;
-  ( match config.hashes with
-    | [] when v_max >= TLS_1_2                          ->
-       invalid "TLS 1.2 configured but no hashes provided"
-    | hs when not (List_set.subset hs supported_hashes) ->
-       invalid "Some hash algorithms are not supported"
-    | _                                                 ->
-       () ) ;
+  ( match config.signature_algorithms with
+    | [] when v_max >= TLS_1_2 ->
+       invalid "TLS 1.2 configured but no signature algorithms provided"
+    | hs when not (List_set.subset hs supported_signature_algorithms) ->
+       invalid "Some signature algorithms are not supported"
```

Comment:
As a user it would probably be nice to be told *which* algorithms are not supported (and the TLS version) so you can remove them from your config?

Reply:
to me this sounds like a general potential improvement to the configuration validation, which should be done -- but best be done as a separate improvement to avoid this PR increasing endlessly in scope and size.